src/sch/Sch.jl

```diff
@@ -747,30 +650,61 @@ function schedule!(ctx, state, procs=procs_to_use(ctx))
 
         # Fire all newly-scheduled tasks
         @label fire_tasks
-        @sync for gpp in keys(to_fire)
-            @async fire_tasks!(ctx, to_fire[gpp], gpp, state)
+        for gpp in keys(to_fire)
+            fire_tasks!(ctx, to_fire[gpp], gpp, state)
         end
+
         append!(state.ready, failed_scheduling)
     end
 end
 
-# Main responsibility of this function is to check if new procs have been pushed to the context
-function assign_new_procs!(ctx, state, assignedprocs)
-    ps = procs_to_use(ctx)
-    # Must track individual procs to handle the case when procs are removed
-    diffps = setdiff(ps, assignedprocs)
-    if !isempty(diffps)
+"""
+Monitors for workers being added/removed to/from `ctx`, sets up or tears down
+per-worker state, and notifies the scheduler so that work can be reassigned.
+"""
+function monitor_procs_changed!(ctx, state)
+    # Load current set of procs
+    old_ps = procs_to_use(ctx)
+
+    while !state.halt.set
+        # Wait for the notification that procs have changed
+        lock(ctx.proc_notify) do
+            wait(ctx.proc_notify)
```

Comment:
Will this sit here and wait (potentially) forever after compute_dag completes? I don't know what is the consequence of a task dangling around, but I suppose it prevents memory from being GC:d.

Perhaps notify at the end of compute_dag, maybe even inside some `finally`?

Reply:
Ah, good catch, I forgot about this! Yes, will add something like that.